samples/compiler_plugins/xnnpack/src/xnnpack/Conversion/StablehloToXnnpack.cpp

```diff
@@ -14,20 +17,144 @@
+static Value getUnconvertedValue(Value val) {
+  if (auto convert = val.getDefiningOp<mlir::stablehlo::ConvertOp>())
+    return getUnconvertedValue(convert.getOperand());
+  return val;
+}
+
+// Go down the use-def chain of `stablehlo.convert` ops starting at `val`, and
+// get the result of the last convert op in the chain.
+static Value getConvertedValue(Value val) {
+  if (val.hasOneUse()) {
+    if (auto convert =
+            dyn_cast<stablehlo::ConvertOp>(*val.getUsers().begin())) {
+      return getConvertedValue(convert.getResult());
+    }
+  }
+  return val;
+}
+
 namespace {
-class ConvertDotGeneralOp
+// A fully connected layer is a sequence of stablehlo ops that behave as the
+// computation `matmul(input, weight.T()) + bias`.
+//
+// See `ConvertFullyConnectedLayer::getFullyConnectedInfo` for the constraints
+// on the fully connected layer pattern.
+//
+// TODO: Add support for `bias`
+class ConvertFullyConnectedLayer
     : public OpConversionPattern<mlir::stablehlo::DotGeneralOp> {
  public:
   using OpConversionPattern::OpConversionPattern;
+  static bool isFullyConnectedLayer(mlir::stablehlo::DotGeneralOp op) {
+    auto error = [](std::string _) { return failure(); };
+    return succeeded(
+        ConvertFullyConnectedLayer::getFullyConnectedInfo(op, error));
+  }
+
   LogicalResult matchAndRewrite(
       mlir::stablehlo::DotGeneralOp op, OpAdaptor adaptor,
       ConversionPatternRewriter &rewriter) const override {
-    Value lhs = adaptor.getLhs();
-    Value rhs = adaptor.getRhs();
-    rewriter.replaceOpWithNewOp<Xnnpack::BatchMatrixMultiplyOp>(
-        op, op.getType(), lhs, rhs);
+    auto error = [op, &rewriter](std::string msg) {
+      return rewriter.notifyMatchFailure(op, msg);
+    };
+    FailureOr<FullyConnectedInfo> maybeInfo = getFullyConnectedInfo(op, error);
+    if (failed(maybeInfo)) {
+      return error("unable to get fully connected info");
+    }
+    auto info = *maybeInfo;
+
+    auto inputType = info.input.getType().cast<RankedTensorType>();
+    auto weightType = info.weight.getType().cast<RankedTensorType>();
+    auto outputType = info.output.getType().cast<RankedTensorType>();
+    if (!inputType.getElementType().isInteger(8) ||
+        !weightType.getElementType().isInteger(4) ||
+        !outputType.getElementType().isF32()) {
+      return error(
+          "unimplemented: fully connected layer without i8 input, i4 weight, "
+          "or f32 output");
+    }
+
+    if (info.needsTranspose) {
+      int64_t weightRank = weightType.getRank();
+      SmallVector<int64_t> perm(llvm::to_vector(llvm::seq(weightRank)));
+      std::swap(perm[perm.size() - 1], perm[perm.size() - 2]);
+      DenseIntElementsAttr permAttr = rewriter.getI64TensorAttr(perm);
+      info.weight = rewriter.create<stablehlo::TransposeOp>(
+          op.getLoc(), info.weight, permAttr);
+    }
+
+    Operation *outputDefiningOp = info.output.getDefiningOp();
+    if (outputDefiningOp != op) rewriter.eraseOp(op);
```

Comment:
`op` should be DCE'ed after `outputDefiningOp` is replaced, so this shouldn't be needed. Also, the code seems to assume that there is only one convert, but the code above checks if there is a convert recursively.

Reply:
This is a [requirement](https://mlir.llvm.org/docs/PatternRewriter/#restrictions) of the `matchAndRewrite` method:

> The root operation is required to either be: updated in-place, replaced, or erased.

Also, there is no assumption that there is only one convert. The very last convert is the one that gets replaced with the xnnpack op. This leaves any use-def chain of extra `convert`s as dead code, which will get DCEd.

The check here is to prevent erasing the op that will get replaced with the xnnpack.

Reply:
I think I see what you mean. If there is a chain `dot_general -> convert -> convert`, and I delete the `dot_general` and replace the last convert, the middle convert will be in an invalid state because its operand no longer exists.

Reply:
Ah, good to know. I was wondering also whether we want to match the convert from i32 to f32 as the root node for the case too, which makes the last node in the tree as the root node. (If the meaning of the root node is really the root node of the tree. ;-))

Reply:
I thought about that. My reasoning for choosing to match on the `dot_general` op is that the `dot_general` represents the _core_ computation of the fully connected layer, whereas the `convert`s are just details specific to the flavor of fully connected layer we are generating. In other words, each pattern in the set of potential fully connected layer patterns will contain a `dot_general` plus other ops that will change from pattern to pattern. Thus, matching on `dot_general` makes the pattern resilient to small differences in the way the fully connected layers are expressed in the model. For example, in some cases there might be no `convert`s after the `dot_general` because the `dot_general` properly encodes the right result type and does the casting inside. In this case, nothing would have to change about our pattern except creating the right `xnnpack` op for it.

In any case, the fix for the erasure of ops is fairly simple. Fixed here: https://github.com/ramiro050/iree/pull/12/commits/d5a79fb9ac20d92dc733c42ae0d19b2aa7077432

Reply:
> I think I see what you mean. If there is a chain `dot_general -> convert -> convert`, and I delete the `dot_general` and replace the last convert, the middle convert will be in an invalid state because its operand no longer exists.

Usually, the multiple converts can be converted into a single one by a peephole optimization if it is not handled by canonicalization, so just assuming there is a preparation pass before it and making the current rewriter pass as simple as possible would be a better approach.